src/petals/server/memory_cache.py

```diff
@@ -84,41 +90,67 @@ async def allocate_cache(self, *descriptors: TensorDescriptor) -> AsyncContextMa
         """
+        try:
+            async with self._wait_for_free_memory(alloc_size, timeout):
+                async with enter_asynchronously(self._lock_metadata):
+                    handles = tuple(int(self.handle_counter) + i for i in range(len(descriptors)))
+                    self.current_size_bytes += alloc_size
+                    self.handle_counter += len(handles)  # note: this will eventually overflow and it is okay
+                    self._pipe_send.send((handles, descriptors))
+                    return handles
+        except TimeoutError:
+            raise AllocationFailed(f"Could not allocate {alloc_size} (timeout={timeout})")
 
+    @contextlib.asynccontextmanager
+    async def _wait_for_free_memory(self, alloc_size: int, timeout: Optional[float]):
+        start_time = time.perf_counter()
         loop = asyncio.get_event_loop()
-        async with hivemind.utils.enter_asynchronously(self._lock_acquire_memory):
+        if timeout == 0:  # if waiting is not allowed, fail when you or anyone else begins waiting
+            stop_when_completes = loop.run_in_executor(None, self._cache_overfull_event.wait)
+        else:  # otherwise, only fail if timeout expires
+            stop_when_completes = asyncio.sleep(timeout if timeout is not None else float("inf"))
+
+        async with wait_for_aenter(enter_asynchronously(self._lock_acquire_memory), stop_when_completes):
             if self.current_size_bytes + alloc_size > self.max_size_bytes:
-                await loop.run_in_executor(None, self._wait_until_available, alloc_size, self.alloc_timeout)
-            with self._lock_metadata:
-                handles = tuple(int(self.handle_counter) + i for i in range(len(descriptors)))
-                self.current_size_bytes += alloc_size
-                self.handle_counter += len(handles)  # note: this will eventually overflow and it is okay
-                self._pipe_send.send((handles, descriptors))
-                return handles
-
-    def _free(self, alloc_size: int, alloc_task: asyncio.Task) -> None:
+                try:
+                    self._cache_overfull_event.set()
+                    elapsed_time = time.perf_counter() - start_time
+                    remaining_timeout = max(0.0, timeout - elapsed_time) if timeout is not None else None
+                    await loop.run_in_executor(None, self._wait_until_available, alloc_size, remaining_timeout)
+                finally:
+                    self._cache_overfull_event.clear()
+            yield
+
+    async def _schedule_free(self, alloc_size: int, alloc_task: asyncio.Task):
+        """
+        This method should be called inside asyncio.shield() because:
+            - hivemind.utils.enter_asynchronously() does not always release the lock on cancellation
+            - _schedule_free() must finish freeing memory even in case of cancellation
+        """
```

Comment:
Please revert this change - this function cannot `await` and **must** return instantly (see comments above).

```suggestion
    def _free(self, alloc_size: int, alloc_task: asyncio.Task):
```